### ISSUE TITLE
Handle  dconnect failed error: 10038

### ISSUE THREAD

**Issue by RmanSC** [JamesBremner/in2out#7]
Tested with in2out v1.2.2 and previous version of raven-set (not 1.05)

Test was performed with:

Long message with /n at the end was sent from separate package sender app through 5000 port. It is similar to real job packages.
i2out.exe output was sent to 127.0.0.1:5001
Listener app is off.
With listener app off in2out.exe stopped working after some time.

Following message is sent over 5000 port:

message to sent.txt

Here are response:

Attempt1:

attempt 1.txt

Then in2out stopped working.

Attemp 2:

attempt 1.txt

Then in2out.exe stopped working.

**Comment by RmanSC** [comment]
Published in [RmanSC-patch-1](https://github.com/JamesBremner/in2out/tree/RmanSC-patch-1)

**Comment by RmanSC** [comment]
Tested with 1.05 Raven-set. Same situation. Will report more in details later.

**Comment by JamesBremner** [comment]
It looks like you wanted to attach some files.  They are missing.  You can attach files by dragging and dropping, selecting or pasting them

Like this, a screenshot pasted from screenshot tool

![image](https://user-images.githubusercontent.com/2046227/185787750-2461b828-8496-4dd6-bed1-85b010b330d7.png)

**Comment by RmanSC** [comment]
[21082022 attempt 1.txt](https://github.com/JamesBremner/in2out/files/9388926/21082022.attempt.1.txt)
[21082022 attempt 2.txt](https://github.com/JamesBremner/in2out/files/9388929/21082022.attempt.2.txt)
[21082022 message to sent.txt](https://github.com/JamesBremner/in2out/files/9388930/21082022.message.to.sent.txt)

**Comment by JamesBremner** [comment]
What does "stopped working" mean?  Does it mean:

- incorrect output
- quit with error message
- quit with no message
- hung

What does "after some time." mean?  Seconds? Minutes? hours? Some number of input lines?

**Comment by RmanSC** [comment]
In attempt 1 quit with no message
In attempt 2 quit with error message

**Comment by JamesBremner** [comment]
This seems to have something to do with "package"  This is a term you started using without any definition.  There was no mention of :packages" in your bid request nor in our discussions about requirements.  

When you first mentioned a package I opened an issue ( #9 ) asking you about this.  I assigned it to you.  That was 9 days ago.  I am still waiting for an answer.

**Comment by JamesBremner** [comment]
```
07/04/2016 11:41:16     1       1       0.0
07/04/2016 11:41:17     1       1       0.0
07/04/2016 11:41:18     1    07/04/2016 11:41:16     1       1       0.0
07/04/2016 11:41:17     1       1       0.0
07/04/2016 11:41:18     1       1       0.0
07/04/2016 11:41:19     1       1       0.0
07/04/☺PжП07/04/2016 11:41:16     1       1       0.0
07/04/2016 11:41:17     1       1       0.0
07/04/2016 11:41:18     1       1       0.0
07/04/2016 11:41:19     1       1       0.0
07/04/☺PжПрt
erminate called after throwing an instance of 'Istd::runtime_errornput Hex: '3
0   what():  dconnect failed error: 10038
a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 34 35 9 30 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 34 36 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 34 37 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 34 38 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 34 39 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 35 30 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 35 31 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 35 32 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 35 33 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 35 34 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 35 35 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 35 36 9 30 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 35 37 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 35 38 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 35 39 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 31 3a 30 30 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 31 3a 30 31 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 31 3a 30 32 9 30 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 31 3a 30 33 9рt
erminate called after throwing an instance of 'Istd::runtime_errornput Hex: '3
0   what():  dconnect failed error: 10038
a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 34 35 9 30 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 34 36 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 34 37 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 34 38 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 34 39 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 35 30 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 35 31 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 35 32 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 35 33 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 35 34 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 35 35 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 35 36 9 30 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 35 37 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 35 38 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 35 39 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 31 3a 30 30 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 31 3a 30 31 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 31 3a 30 32 9 30 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 31 3a 30 33 9   1       0.0
07/04/2016 11:41:19     1       1       0.0
07/04/☺PжПрt
erminate called after throwing an instance of 'Istd::runtime_errornput Hex: '3
0   what():  dconnect failed error: 10038
a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 34 35 9 30 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 34 36 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 34 37 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 34 38 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 34 39 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 35 30 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 35 31 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 35 32 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 35 33 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 35 34 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 35 35 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 35 36 9 30 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 35 37 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 35 38 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 30 3a 35 39 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 31 3a 30 30 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 31 3a 30 31 9 31 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 31 3a 30 32 9 30 9 31 9 30 2e 30 d a 30 37 2f 30 34 2f 32 30 31 36 20 31 31 3a 34 31 3a 30 33 9
```

This is the last data received.

`07/04/☺PжПрt`

It looks like the data received was corrupted.

**Comment by RmanSC** [comment]
Yes, but from sender point it was sent in full.

**Comment by JamesBremner** [comment]
I think the corrupt data is a coincidence.  It may be the result of a buffer race in the output stream between the input handler and the exception handler.

The real issue is that  `dconnect failed error: 10038` is not being handled correctly.  I will look into this.

**Comment by JamesBremner** [comment]
The winsock2 connect() method is suppose to return an error number if anything goes wrong.  This usually happens OK.  In this case it seem to have thrown an exception instead.  I will add code to handle this.

**Comment by JamesBremner** [comment]
v1.2.3

**Comment by RmanSC** [comment]
Program stop with error message "cIn2Out::run: send on invalid socket"

See test from current build:
When you don't turn on the listener:
[22082022 attemp 1.txt](https://github.com/JamesBremner/in2out/files/9395392/22082022.attemp.1.txt)

When you turn on and then turn off the listener:
[22082022 attemp 3.txt](https://github.com/JamesBremner/in2out/files/9395394/22082022.attemp.3.txt)

**Comment by JamesBremner** [comment]
The input is receiving garbles

```
07/04/2016 11:38:23     1       1       0.0
07/04/2016 11:38:24     0       1       0☺@Q"┬
```

**Comment by JamesBremner** [comment]
This is unexpected

`mm/dd/yyyy hh:mm:ss     psi     psi     bbl/min`

I will open a new issue for this.

**Comment by RmanSC** [comment]
Test was successful. (copied from issue #9 
Test package (message), that was sent:
[21082022 message to sent.txt](https://github.com/JamesBremner/in2out/files/9403950/21082022.message.to.sent.txt)

Newest version 1.4.3 was tested on:

receiving data for 5 min without listener - success
receiving data for 5 min with listener - success
receiving data for 5 min after disconnection from listener - success
Only case, that make the app terminate with bad_alloc message was when I tried to turn on/off listener quickly to simulate bad connection. It happen to me only once. I tried to write better log file, but couldn't trigger that mistake and log of that occurance was erased already.
I think, it's minor issue and this issue can be closed.
